### PR TITLE
Cleanup processes started by UseReactDevelopmentServer and UseAngularCliServer

### LINKED ISSUE #11597: UseReactDevelopmentServer with UseSpa keeps console open.
### Describe the bug
When I use "UseReactDevelopmentServer", the title of the console changes to "npm" and when I stop debugging with Visual Studio, this console keeps open. This can be pretty annoying when debugging fast.

### To Reproduce
**Version:** 2.2.300, 3.0.0-preview6
1. Create a React app using the template (or use UseReactDevelopmentServer).
2. Start Visual Studio debugging or use dotnet run.
3. Stop debugging.

### Expected behavior
When I stop, the console should be closed.

### Additional context
Looking at the code, I don't see any code to close npm. It could use IApplicationLifetime to close the process on Shutdown, for example.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #11597

### CODE REVIEW

File: src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs [hunk 1]
```diff
@@ -64,7 +67,7 @@ internal static class AngularCliMiddleware
         }
 
         private static async Task<AngularCliServerInfo> StartAngularCliServerAsync(
-            string sourcePath, string scriptName, string pkgManagerCommand, int portNumber, ILogger logger)
+            string sourcePath, string scriptName, string pkgManagerCommand, int portNumber, ILogger logger, CancellationToken applicationStoppingToken)
```
Comment (ryanbrandenburg) [inline src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs]: This design seems reasonable and good to me, but I don't have a background with `IHostApplicationLifetime`, @SteveSandersonMS would you know if there are any pitfalls here, or know who might know?

**Review by ryanbrandenburg — CHANGES_REQUESTED** [review]
As far as I can tell the actual code change here is good, but I believe the tests as-is are likely to be flaky. Should be in a much better place once that's resolved.

File: src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs]: `IHostApplicationLifetime` sounds like a totally reasonable thing to use for this. AFAIK it's the primary thing we have that signals when an application is shutting down.

It's not absolutely bulletproof (if the application dies due to an unhandled exception or being forcibly killed then I don't think it's guaranteed to notify) but this should still be an improvement.

File: src/Middleware/SpaServices.Extensions/test/Microsoft.AspNetCore.SpaServices.Extensions.Tests.csproj [hunk 2]
```diff
@@ -14,4 +14,10 @@
     <Content Include="js\**\*" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Update="package.json">
```
[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/Middleware/SpaServices.Extensions/test/Microsoft.AspNetCore.SpaServices.Extensions.Tests.csproj]: npm exits immediately

File: src/Middleware/SpaServices.Extensions/test/SpaServicesExtensionsTests.cs [hunk 3]
```diff
@@ -24,9 +39,62 @@ public void UseSpa_ThrowsInvalidOperationException_IfRootpathNotSet()
             Assert.Equal("No RootPath was set on the SpaStaticFilesOptions.", exception.Message);
         }
 
+        [ConditionalFact]
+        public async Task UseSpa_KillsRds_WhenAppIsStopped()
```
[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/Middleware/SpaServices.Extensions/test/SpaServicesExtensionsTests.cs]: Yes 😄

File: src/Middleware/SpaServices.Extensions/test/SpaServicesExtensionsTests.cs [hunk 4]
```diff
@@ -24,9 +39,62 @@ public void UseSpa_ThrowsInvalidOperationException_IfRootpathNotSet()
...
+            // Give node a moment to start up
+            await Task.Delay(TimeSpan.FromSeconds(3));
+            var processCount = Process.GetProcesses().Length;
+
+            // Act
+            applicationLifetime.StopApplication();
+
+            // Assert
+            AssertNoErrors();
+            Assert.True(Process.GetProcesses().Length < processCount, $"Expected less than {processCount} processes");
```
[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/Middleware/SpaServices.Extensions/test/SpaServicesExtensionsTests.cs]: Added a `DiagnosticSource` event with the process id

File: src/Middleware/SpaServices.Extensions/test/SpaServicesExtensionsTests.cs [hunk 5]
```diff
@@ -24,9 +39,62 @@ public void UseSpa_ThrowsInvalidOperationException_IfRootpathNotSet()
...
+        }
+
+        private string GetCommand()
+            => RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "waitWindows" : "wait";
```
[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/Middleware/SpaServices.Extensions/test/SpaServicesExtensionsTests.cs]: The command must run indefinitely until the process is killed

**Comment by mkArtakMSFT** [comment]
@ryanbrandenburg are we waiting for more changes from @AndriySvyryd here?

File: src/Middleware/SpaServices.Extensions/test/Microsoft.AspNetCore.SpaServices.Extensions.Tests.csproj [hunk 6]
```diff
@@ -1,17 +1,26 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
     <TestDependsOnNode>true</TestDependsOnNode>
+	<!-- Depends on npm which is not picked up on helix -->
+	<BuildHelixPayload>false</BuildHelixPayload>
```
Comment (ryanbrandenburg) [inline src/Middleware/SpaServices.Extensions/test/Microsoft.AspNetCore.SpaServices.Extensions.Tests.csproj]: Fair, but let's follow the pattern [established elsewhere](https://github.com/dotnet/aspnetcore/blob/0517745c08d1bc85009f1ca41a41d77176eeea60/src/Hosting/test/FunctionalTests/Microsoft.AspNetCore.Hosting.FunctionalTests.csproj#L7) and include a comment with the GitHub issue tracking the fact that Helix doesn't run this ATM.

**Review by ryanbrandenburg — APPROVED** [review]
Awesome! I really like the diagnostic listener pattern, I might end up stealing that in some spots.

One small comment inclusion then I think we can call this good.

**Comment by leandroaraujosouza** [comment]
Hi @AndriySvyryd can you tell when this is gonna be published with the SDK ?

**Comment by AndriySvyryd** [comment]
Nightly builds should already have this. Preview1 will be out soon*. The RTM version will be out in November.

**Comment by JSLGeeganage** [comment]
Hi @AndriySvyryd, the related issue still can produce in dot net version 5.0.100. If I start debug using (not using IIS) Visual Studio (16.8.2), node process still open even I stopped the debug.

But the _crt+c_ will kill all the node process when you use _dotnet run_ command using developer user command prompt.

**Pull request merged**